### ISSUE TITLE
Training with Large Dataset Causes Infinite Stall

### ISSUE THREAD

**Issue by fishbotics** [PyTorchLightning/pytorch-lightning#13126]
## 🐛 Bug
I am training training two jobs that use the exact same code but different h5py files (same keys) to represent the data. One of the h5py files leads to a dataset that has 650,000,000 data points in the training set and one has 150,000,000. The training job with the larger dataset seems to just stop doing anything after a number of hours. After some number of hours, the CPU activity seems to die down. Neither job is out of memory (either system or GPU) but the larger job just seems to... stop working (although it doesn't crash).

Any ideas what might be going on? Both are still running and I can provide any info that might be helpful.

PyTorch Lightning Version (e.g., 1.5.0): 1.6.3
PyTorch Version (e.g., 1.10): 1.11
Python version (e.g., 3.9): 3.8
OS (e.g., Linux): Ubuntu 20.04
CUDA/cuDNN version: 11.3
GPU models and configuration: 8 A100s using ddp
How you installed PyTorch (conda, pip, source): pip
If compiling from source, the output of torch.__config__.show():
Any other relevant information:

**Comment by fishbotics** [comment]
I know this isn't exactly helpful, but I finally killed the job and it seems like it's some threading race condition thing. When I kill the process, the traceback looks like all of the DDP processes were waiting on a lock

```
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/process.py", line 318, in _bootstrap
    util._exit_function()
  File "/usr/lib/python3.8/multiprocessing/util.py", line 360, in _exit_function
    _run_finalizers()
  File "/usr/lib/python3.8/multiprocessing/util.py", line 300, in _run_finalizers
    finalizer()
  File "/usr/lib/python3.8/multiprocessing/util.py", line 224, in __call__
    res = self._callback(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 195, in _finalize_join
    thread.join()
  File "/usr/lib/python3.8/threading.py", line 1011, in join
    self._wait_for_tstate_lock()
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 1027, in _wait_for_tstate_lock
    elif lock.acquire(block, timeout):
```

Unfortunately, I don't see anything else of interest in there. The GPU utilization was high on several GPUs (and 0 on others), but the memory seemed fine. I saw https://github.com/PyTorchLightning/pytorch-lightning/issues/5725, which looks similar to this issue but my job wasn't out of memory. 

For what it's worth, I tried the fix mentioned in that other thread (move the metrics to cpu) but this does not seem to work with DDP. I get an error saying the tensors need to on CUDA and dense.

**Comment by GGGGGGXY** [comment]
same issue.
when GPU utilization is 100%, but the training is hanging. Need ctrl+c to kill the training and start again.

<img width="1609" alt="image" src="https://user-images.githubusercontent.com/15215819/169766944-d12be036-83df-481e-85cd-ce22d86e250a.png">

**Comment by carmocca** [comment]
Unfortunately, we would need more information and ideally a reproducible script to check what could be the issue.

**Comment by angadkalra** [comment]
I'm having the exact same issue. For me, it looks like with a large dataset (500 GBs, 3000 data points), the machine free memory gets very low and it's causing some sort of process hang. The machine has total memory 256 GBs, the buff/cache during training is 230gb, used is 19gb, and free is around 4gb but goes up and down a lot. The data files are stored in EFS and mounted on the machine (same region). 

In the middle of training, training will freeze in the middle of an epoch and when I check whats going on with `top i`, theres only a few python processes left (5 when using num_workers = 4) and 3 out of 4 GPUs are at 100%, the other gpu is at 0%. I barely get to 20 epochs before I have to hit ctrl+z, kill all python processes, and resume training from last.ckpt. CTRL+C doesn't work most of the time to kill training.

When I run training with 1000 data points (160 GBs), theres 120gb free memory the entire time, buff/cache is at 105gb and used is around 20gb. Training never stops and runs at 2it/s the entire time. I'm able to run 150 epochs in a row.

**Comment by angadkalra** [comment]
@carmocca Why does the machine memory buff/cache increase with dataset size? What are we storing across all epochs?

**Comment by angadkalra** [comment]
@Borda can you please help with this?

**Comment by Borda** [comment]
> @Borda can you please help with this?

could you share your code to reproduce?

**Comment by angadkalra** [comment]
> > @Borda can you please help with this?
> 
> could you share your code to reproduce?

I'll work on creating a colab with the code I'm using. 

However, I think it's the dataset size thats causing my issue. Using the exact same training code with 1/3 of the data gives great performance and no freezing. Could you recommend any links/training tips/documentation that can help with reducing the amount of buff/cache that is used during training?

**Comment by angadkalra** [comment]
This is what happens to me, been stuck here for an hour.

![Screen Shot 2022-07-14 at 1 47 58 PM](https://user-images.githubusercontent.com/8902328/179082712-ca2784bf-47cb-4bba-a60d-2d458eda75c5.png)

**Comment by akihironitta** [comment]
> Could you recommend any links/training tips/documentation that can help with reducing the amount of buff/cache that is used during training?

Hi @angadkalra, here's a list of possible causes of your performance issue #12398. Also, I'd suggest skimming through [the existing issues labelled "performance"](https://github.com/Lightning-AI/lightning/issues?q=is%3Aissue+label%3Aperformance). Let me know if any of these work. If the issue still persists, please share your code and env details so that I can investigate it.

**Comment by angadkalra** [comment]
> > Could you recommend any links/training tips/documentation that can help with reducing the amount of buff/cache that is used during training?
> 
> Hi @angadkalra, here's a list of possible causes of your performance issue #12398. Also, I'd suggest skimming through [the existing issues labelled "performance"](https://github.com/Lightning-AI/lightning/issues?q=is%3Aissue+label%3Aperformance). Let me know if any of these work. If the issue still persists, please share your code and env details so that I can investigate it.

None of those tips worked. I switched to a machine with more memory and now my training runs for 100 epochs at 2.5it/s consistently per epoch without stopping once.

**Comment by angadkalra** [comment]
@akihironitta what is the majority of items in buff/cache between epochs?

**Comment by angadkalra** [comment]
@fishbotics can you try increasing your system memory for the larger dataset? I'm training on cloud VMs, so its easy for me to provision a machine with more memory...don't know your situation, but that could be the solution. Also I switched back to PL 1.5.4, torch 1.9.0+cu111.

**Comment by fishbotics** [comment]
I can't increase the system memory since I'm already training with the largest node available to me (something like 400gb of memory). When I looked in htop, I did not see that the job was out of memory.

@angadkalra with your job, was the system previously reporting that it was out of memory? Or did you just happen to change it and it was fixed?

**Comment by angadkalra** [comment]
> I can't increase the system memory since I'm already training with the largest node available to me (something like 400gb of memory). When I looked in htop, I did not see that the job was out of memory.
> 
> @angadkalra with your job, was the system previously reporting that it was out of memory? Or did you just happen to change it and it was fixed?

So I was watching the epoch with "top i" and I was watching how much free mem was left when the job froze, and everytime it was like 3 gb or 1 gb, something less than 5. And my colleague suggested it could be something to do with transferring objects to disk because of memory running low, so I upped the memory and now it runs perfectly and free mem is always 15 gb and higher. 

I'm making sure I'm using tensors and numpy arrays in all my training code and datamodule code, because I know python lists end up making memory spike. Also seed_everything helps with stable training (saw it another issue). I'm logging two metrics, loss and AUROC, for both train and val (using torchmetrics). Pretty basic setup and code.

**Comment by akihironitta** [comment]
@angadkalra Sorry for the delay.

> @akihironitta what is the majority of items in buff/cache between epochs?

What I've seen often in users' scripts is that they use `training_epoch_end` which holds all the return values from the `training_step`.

> Also I switched back to PL 1.5.4, torch 1.9.0+cu111.

Do you see some difference between PL 1.5.4 and 1.6.x? If there's any, that'll be something that we should look into, but without any script to reproduce the behaviour, we can't really do anything other than mentioning possible causes listed in #12398. Is there any chance you could provide us with your complete script or a minimal script based your script?

Also, would you mind also trying the most recent stable release, PL 1.7?

**Comment by angadkalra** [comment]
I've upgraded to PL 1.7.0 and latest pytorch + cu113. Everything works flawlessly when the machine has ample memory free (my machine has 70gb unused as training goes on). When my dataset got larger, I had to use more GPUs (8 instead of 4) because that allowed my machine to get more memory (GCP requirement).

**Comment by anishazaveri** [comment]
Same issue. Long training job (~900k steps/epoch). Training stalls mid-epoch (somewhere around ~300k steps) without any error message. This corresponds to a constant increase in buff/cache usage in the CPU until it's full. 

Lightning version: 1.7.0
Torch version: 1.10.0
CUDA: 11.5

I solved it by making num_workers=0. This still leads to an increase in buff/cache but it levels off before it is full.

**Comment by fishbotics** [comment]
Just coming back to this. I am still seeing this issue when the GPU Vol Util (NVIDIA A100s) creep up to 100%. My current strategy is to just kill the job and restart from a checkpoint, but that's less than ideal.

**Comment by fishbotics** [comment]
Also, FWIW this is what my GPU utilization chart looks like when it stalls:

![image](https://user-images.githubusercontent.com/45679966/211162428-38c4ce18-ddcc-48a7-a639-d0cbe0ff2fa0.png)

**Comment by carmocca** [comment]
When you kill it, is the stacktrace still the same as https://github.com/Lightning-AI/lightning/issues/13126#issuecomment-1134201225?

**Comment by fishbotics** [comment]
Yes, it's more or less still the same. I'll include the whole thing here just in case there's some extra info I missed.

`Traceback (most recent call last):                                                                                                                                                                                                                     
  File "/usr/lib/python3.8/multiprocessing/process.py", line 318, in _bootstrap                                                                                                                                                                        
    util._exit_function()                                                                                                                                                                                                                              
  File "/usr/lib/python3.8/multiprocessing/util.py", line 360, in _exit_function                                                                                                                                                                       
    _run_finalizers()                                                                                                                                                                                                                                  
  File "/usr/lib/python3.8/multiprocessing/util.py", line 300, in _run_finalizers                                                                                                                                                                      
    finalizer()                                                                                                                                                                                                                                        
  File "/usr/lib/python3.8/multiprocessing/util.py", line 224, in __call__                                                                                                                                                                             
    res = self._callback(*self._args, **self._kwargs)                                                                                                                                                                                                  
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 195, in _finalize_join                                                                                                                                                                     
    thread.join()                                                                                                                                                                                                                                      
  File "/usr/lib/python3.8/threading.py", line 1011, in join                                                                                                                                                                                           
    self._wait_for_tstate_lock()                                                                                                                                                                                                                       
  File "/usr/lib/python3.8/threading.py", line 1027, in _wait_for_tstate_lock                                                                                                                                                                          
    elif lock.acquire(block, timeout):                                                                                                                                                                                                                 
KeyboardInterrupt                                                                                                                                                                                                                                      
Process Process-1494:                                                                                                                                                                                                                                  
Traceback (most recent call last):                                                                                                                                                                                                                     
  File "/usr/lib/python3.8/multiprocessing/process.py", line 318, in _bootstrap                                                                                                                                                                        
    util._exit_function()                                                                                                                                                                                                                              
  File "/usr/lib/python3.8/multiprocessing/util.py", line 360, in _exit_function                                                                                                                                                                       
    _run_finalizers()                                                                                                                                                                                                                                  
  File "/usr/lib/python3.8/multiprocessing/util.py", line 300, in _run_finalizers                                                                                                                                                                      
    finalizer()                                                                                                                                                                                                                                        
  File "/usr/lib/python3.8/multiprocessing/util.py", line 224, in __call__                                                                                                                                                                             
    res = self._callback(*self._args, **self._kwargs)                                                                                                                                                                                                  
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 195, in _finalize_join                                                                                                                                                                     
    thread.join()                                                                                                                                                                                                                                      
  File "/usr/lib/python3.8/threading.py", line 1011, in join                                                                                                                                                                                           
    self._wait_for_tstate_lock()                                                                                                                                                                                                                       
  File "/usr/lib/python3.8/threading.py", line 1027, in _wait_for_tstate_lock                                                                                                                                                                          
    elif lock.acquire(block, timeout):                                                                                                                                                                                                                 
KeyboardInterrupt                                                                                                                                                                                                                                      
Process Process-1492:                                                                                                                                                                                                                                  
Traceback (most recent call last):                                                                                                                                                                                                                     
  File "/usr/lib/python3.8/multiprocessing/process.py", line 318, in _bootstrap                                                                                                                                                                        
    util._exit_function()                                    
  File "/usr/lib/python3.8/multiprocessing/util.py", line 360, in _exit_function
    _run_finalizers()                                        
  File "/usr/lib/python3.8/multiprocessing/util.py", line 300, in _run_finalizers
    finalizer()                                              
  File "/usr/lib/python3.8/multiprocessing/util.py", line 224, in __call__
    res = self._callback(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 195, in _finalize_join
    thread.join()                                            
  File "/usr/lib/python3.8/threading.py", line 1011, in join
    self._wait_for_tstate_lock()
  File "/usr/lib/python3.8/threading.py", line 1027, in _wait_for_tstate_lock
    elif lock.acquire(block, timeout):
KeyboardInterrupt                                            
Process Process-1487:                                        
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/process.py", line 318, in _bootstrap
    util._exit_function()                                    
  File "/usr/lib/python3.8/multiprocessing/util.py", line 360, in _exit_function
    _run_finalizers()                                                                                                                                                                                                                                  
  File "/usr/lib/python3.8/multiprocessing/util.py", line 300, in _run_finalizers                                                                                                                                                                      
    finalizer()                                                                                                                                                                                                                                        
  File "/usr/lib/python3.8/multiprocessing/util.py", line 224, in __call__                                                                                                                                                                             
    res = self._callback(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 195, in _finalize_join
    thread.join()                                            
  File "/usr/lib/python3.8/threading.py", line 1011, in join
    self._wait_for_tstate_lock()
  File "/usr/lib/python3.8/threading.py", line 1027, in _wait_for_tstate_lock
    elif lock.acquire(block, timeout):
KeyboardInterrupt                                            
Process Process-1497:                                        
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/process.py", line 318, in _bootstrap
    util._exit_function()                                    
  File "/usr/lib/python3.8/multiprocessing/util.py", line 360, in _exit_function
    _run_finalizers()                                        
Process Process-1490:                                        
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/process.py", line 318, in _bootstrap
    util._exit_function()                                    
  File "/usr/lib/python3.8/multiprocessing/util.py", line 360, in _exit_function
    _run_finalizers()                                        
  File "/usr/lib/python3.8/multiprocessing/util.py", line 300, in _run_finalizers
    finalizer()                                              
  File "/usr/lib/python3.8/multiprocessing/util.py", line 224, in __call__
    res = self._callback(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 195, in _finalize_join
    thread.join()                                            
  File "/usr/lib/python3.8/threading.py", line 1011, in join
    self._wait_for_tstate_lock()
  File "/usr/lib/python3.8/threading.py", line 1027, in _wait_for_tstate_lock
    elif lock.acquire(block, timeout):
KeyboardInterrupt                                            
Process Process-1481:                                        
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/process.py", line 318, in _bootstrap
    util._exit_function()                                    
  File "/usr/lib/python3.8/multiprocessing/util.py", line 360, in _exit_function
    _run_finalizers()                                        
  File "/usr/lib/python3.8/multiprocessing/util.py", line 300, in _run_finalizers
    finalizer()                                              
  File "/usr/lib/python3.8/multiprocessing/util.py", line 224, in __call__
    res = self._callback(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 195, in _finalize_join
    thread.join()                                            
  File "/usr/lib/python3.8/threading.py", line 1011, in join
    self._wait_for_tstate_lock()
  File "/usr/lib/python3.8/threading.py", line 1027, in _wait_for_tstate_lock
    elif lock.acquire(block, timeout):
KeyboardInterrupt                                            
Process Process-1489:                                        
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/process.py", line 318, in _bootstrap
    util._exit_function()                                    
  File "/usr/lib/python3.8/multiprocessing/util.py", line 360, in _exit_function
    _run_finalizers()                                        
  File "/usr/lib/python3.8/multiprocessing/util.py", line 300, in _run_finalizers
    finalizer()                                              
  File "/usr/lib/python3.8/multiprocessing/util.py", line 224, in __call__
    res = self._callback(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 195, in _finalize_join
    thread.join()                                            
  File "/usr/lib/python3.8/threading.py", line 1011, in join
    self._wait_for_tstate_lock()
  File "/usr/lib/python3.8/threading.py", line 1027, in _wait_for_tstate_lock
    elif lock.acquire(block, timeout):
KeyboardInterrupt                                            
Process Process-1479:                                        
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/process.py", line 318, in _bootstrap
    util._exit_function()                                    
  File "/usr/lib/python3.8/multiprocessing/util.py", line 360, in _exit_function
    _run_finalizers()                                        
  File "/usr/lib/python3.8/multiprocessing/util.py", line 300, in _run_finalizers
    finalizer()                                              
  File "/usr/lib/python3.8/multiprocessing/util.py", line 224, in __call__
    res = self._callback(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 195, in _finalize_join
    thread.join()                                            
  File "/usr/lib/python3.8/threading.py", line 1011, in join
    self._wait_for_tstate_lock()
  File "/usr/lib/python3.8/threading.py", line 1027, in _wait_for_tstate_lock
    elif lock.acquire(block, timeout):
KeyboardInterrupt                                            
Process Process-1482:                                        
Process Process-1477:                                        
Process Process-1478:                                        
Process Process-1493:                                        
Traceback (most recent call last):
Process Process-1476:                                        
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/process.py", line 318, in _bootstrap
    util._exit_function()                                    
  File "/usr/lib/python3.8/multiprocessing/util.py", line 360, in _exit_function
    _run_finalizers()                                        
  File "/usr/lib/python3.8/multiprocessing/util.py", line 300, in _run_finalizers
    finalizer()                                              
  File "/usr/lib/python3.8/multiprocessing/util.py", line 224, in __call__
    res = self._callback(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 195, in _finalize_join
    thread.join()                                            
  File "/usr/lib/python3.8/threading.py", line 1011, in join
    self._wait_for_tstate_lock()
  File "/usr/lib/python3.8/threading.py", line 1027, in _wait_for_tstate_lock
    elif lock.acquire(block, timeout):
KeyboardInterrupt                                            
Process Process-1486:                                        
Process Process-1480:                                        
Traceback (most recent call last):
Process Process-1491:                                        
Process Process-1488:                                        
Traceback (most recent call last):
Process Process-1496:                                        
Process Process-1485:                                        
Process Process-1484:                                        
Process Process-1495:                                        
Process Process-1474:                                        
Traceback (most recent call last):
Traceback (most recent call last):
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/process.py", line 318, in _bootstrap
    util._exit_function()                                    
  File "/usr/lib/python3.8/multiprocessing/process.py", line 318, in _bootstrap
    util._exit_function()                                    
  File "/usr/lib/python3.8/multiprocessing/process.py", line 318, in _bootstrap
    util._exit_function()                                    
  File "/usr/lib/python3.8/multiprocessing/util.py", line 360, in _exit_function
    _run_finalizers()                                        
  File "/usr/lib/python3.8/multiprocessing/util.py", line 360, in _exit_function
    _run_finalizers()                                        
  File "/usr/lib/python3.8/multiprocessing/util.py", line 360, in _exit_function
    _run_finalizers()                                        
  File "/usr/lib/python3.8/multiprocessing/util.py", line 300, in _run_finalizers
    finalizer()                                              
  File "/usr/lib/python3.8/multiprocessing/util.py", line 300, in _run_finalizers
    finalizer()                                              
  File "/usr/lib/python3.8/multiprocessing/util.py", line 300, in _run_finalizers
    finalizer()                                              
  File "/usr/lib/python3.8/multiprocessing/util.py", line 224, in __call__
    res = self._callback(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/multiprocessing/util.py", line 224, in __call__
    res = self._callback(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/multiprocessing/util.py", line 224, in __call__
    res = self._callback(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 195, in _finalize_join
    thread.join()                                            
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 195, in _finalize_join
    thread.join()                                            
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 195, in _finalize_join
    thread.join()                                            
  File "/usr/lib/python3.8/threading.py", line 1011, in join
    self._wait_for_tstate_lock()
  File "/usr/lib/python3.8/threading.py", line 1011, in join
    self._wait_for_tstate_lock()
  File "/usr/lib/python3.8/threading.py", line 1011, in join
    self._wait_for_tstate_lock()
  File "/usr/lib/python3.8/threading.py", line 1027, in _wait_for_tstate_lock
    elif lock.acquire(block, timeout):
  File "/usr/lib/python3.8/threading.py", line 1027, in _wait_for_tstate_lock
    elif lock.acquire(block, timeout):
  File "/usr/lib/python3.8/threading.py", line 1027, in _wait_for_tstate_lock
    elif lock.acquire(block, timeout):
KeyboardInterrupt                                            
KeyboardInterrupt                                            
KeyboardInterrupt                                            
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/process.py", line 318, in _bootstrap
    util._exit_function()                                    
  File "/usr/lib/python3.8/multiprocessing/util.py", line 360, in _exit_function
    _run_finalizers()                                        
  File "/usr/lib/python3.8/multiprocessing/util.py", line 300, in _run_finalizers
    finalizer()                                              
  File "/usr/lib/python3.8/multiprocessing/util.py", line 224, in __call__
    res = self._callback(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 195, in _finalize_join
    thread.join()                                            
  File "/usr/lib/python3.8/threading.py", line 1011, in join
    self._wait_for_tstate_lock()                                                                                                                                                                                                             [245/1907]
  File "/usr/lib/python3.8/threading.py", line 1027, in _wait_for_tstate_lock
    elif lock.acquire(block, timeout):
KeyboardInterrupt                                            
  File "/usr/lib/python3.8/multiprocessing/util.py", line 300, in _run_finalizers
    finalizer()                                              
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/util.py", line 224, in __call__
    res = self._callback(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 195, in _finalize_join
    thread.join()                                            
  File "/usr/lib/python3.8/threading.py", line 1011, in join
    self._wait_for_tstate_lock()
  File "/usr/lib/python3.8/threading.py", line 1027, in _wait_for_tstate_lock
    elif lock.acquire(block, timeout):
KeyboardInterrupt                                            
Traceback (most recent call last):
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/process.py", line 318, in _bootstrap
    util._exit_function()                                    
  File "/usr/lib/python3.8/multiprocessing/process.py", line 318, in _bootstrap
    util._exit_function()                                    
  File "/usr/lib/python3.8/multiprocessing/util.py", line 360, in _exit_function
    _run_finalizers()                                        
  File "/usr/lib/python3.8/multiprocessing/util.py", line 360, in _exit_function
    _run_finalizers()                                        
  File "/usr/lib/python3.8/multiprocessing/util.py", line 300, in _run_finalizers
    finalizer()                                              
  File "/usr/lib/python3.8/multiprocessing/util.py", line 300, in _run_finalizers
    finalizer()                                              
  File "/usr/lib/python3.8/multiprocessing/util.py", line 224, in __call__
    res = self._callback(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/multiprocessing/util.py", line 224, in __call__
    res = self._callback(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 195, in _finalize_join
    thread.join()                                            
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 195, in _finalize_join
    thread.join()                                            
  File "/usr/lib/python3.8/threading.py", line 1011, in join
    self._wait_for_tstate_lock()
  File "/usr/lib/python3.8/threading.py", line 1011, in join
    self._wait_for_tstate_lock()
  File "/usr/lib/python3.8/threading.py", line 1027, in _wait_for_tstate_lock
    elif lock.acquire(block, timeout):
  File "/usr/lib/python3.8/threading.py", line 1027, in _wait_for_tstate_lock
    elif lock.acquire(block, timeout):
KeyboardInterrupt                                            
KeyboardInterrupt                                            
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/process.py", line 318, in _bootstrap
    util._exit_function()                                    
  File "/usr/lib/python3.8/multiprocessing/util.py", line 360, in _exit_function
    _run_finalizers()                                        
  File "/usr/lib/python3.8/multiprocessing/util.py", line 300, in _run_finalizers
    finalizer()                                              
  File "/usr/lib/python3.8/multiprocessing/util.py", line 224, in __call__
    res = self._callback(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 195, in _finalize_join
    thread.join()                                            
  File "/usr/lib/python3.8/threading.py", line 1011, in join
    self._wait_for_tstate_lock()
  File "/usr/lib/python3.8/threading.py", line 1027, in _wait_for_tstate_lock
    elif lock.acquire(block, timeout):
KeyboardInterrupt                                            
  File "/usr/lib/python3.8/multiprocessing/process.py", line 318, in _bootstrap
    util._exit_function()                                    
  File "/usr/lib/python3.8/multiprocessing/util.py", line 360, in _exit_function
    _run_finalizers()                                        
  File "/usr/lib/python3.8/multiprocessing/util.py", line 300, in _run_finalizers
    finalizer()                                              
  File "/usr/lib/python3.8/multiprocessing/process.py", line 318, in _bootstrap
    util._exit_function()                                    
  File "/usr/lib/python3.8/multiprocessing/util.py", line 224, in __call__
    res = self._callback(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/multiprocessing/util.py", line 360, in _exit_function
    _run_finalizers()                                        
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 195, in _finalize_join
    thread.join()                                            
  File "/usr/lib/python3.8/multiprocessing/util.py", line 300, in _run_finalizers
    finalizer()                                              
  File "/usr/lib/python3.8/multiprocessing/util.py", line 224, in __call__
    res = self._callback(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/threading.py", line 1011, in join
    self._wait_for_tstate_lock()
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 195, in _finalize_join
    thread.join()                                            
  File "/usr/lib/python3.8/threading.py", line 1011, in join
    self._wait_for_tstate_lock()
  File "/usr/lib/python3.8/threading.py", line 1027, in _wait_for_tstate_lock
    elif lock.acquire(block, timeout):
  File "/usr/lib/python3.8/threading.py", line 1027, in _wait_for_tstate_lock
    elif lock.acquire(block, timeout):
KeyboardInterrupt                                            
KeyboardInterrupt                                            
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/process.py", line 318, in _bootstrap
    util._exit_function()                                    
  File "/usr/lib/python3.8/multiprocessing/util.py", line 360, in _exit_function
    _run_finalizers()                                        
  File "/usr/lib/python3.8/multiprocessing/util.py", line 300, in _run_finalizers
    finalizer()                                              
  File "/usr/lib/python3.8/multiprocessing/util.py", line 224, in __call__
    res = self._callback(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/multiprocessing/queues.py", line 195, in _finalize_join
    thread.join()                                            
  File "/usr/lib/python3.8/threading.py", line 1011, in join
    self._wait_for_tstate_lock()
  File "/usr/lib/python3.8/threading.py", line 1027, in _wait_for_tstate_lock
    elif lock.acquire(block, timeout):
KeyboardInterrupt                                            
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/process.py", line 318, in _bootstrap
    util._exit_function()                                    
  File "/usr/lib/python3.8/multiprocessing/util.py", line 360, in _exit_function
    _run_finalizers()                                        

`

**Comment by AlyaGomaa** [comment]
I had the same issue. after debugging this is what I came up with

From the multiprocessing docs

> By default if a process is not the creator of the queue then on
> exit it will attempt to join the queue’s background thread.

if this is your case, then it is the cause of the deadlock

to avoid it you should call cancel_join_thread() on KeyboardInterrupt
https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Queue.cancel_join_thread
